source/common/stats/thread_local_store.cc

```diff
@@ -14,21 +14,21 @@ namespace Stats {
 ThreadLocalStoreImpl::ThreadLocalStoreImpl(RawStatDataAllocator& alloc)
     : alloc_(alloc), default_scope_(createScope("")),
       tag_producer_(std::make_unique<TagProducerImpl>()),
-      num_last_resort_stats_(default_scope_->counter("stats.overflow")) {}
+      num_last_resort_stats_(default_scope_->counter("stats.overflow")), stats_source_(*this) {}
 
 ThreadLocalStoreImpl::~ThreadLocalStoreImpl() {
   ASSERT(shutting_down_);
   default_scope_.reset();
   ASSERT(scopes_.empty());
 }
 
-std::list<CounterSharedPtr> ThreadLocalStoreImpl::counters() const {
+std::vector<CounterSharedPtr> ThreadLocalStoreImpl::counters() const {
   // Handle de-dup due to overlapping scopes.
-  std::list<CounterSharedPtr> ret;
+  std::vector<CounterSharedPtr> ret;
   std::unordered_set<std::string> names;
   std::unique_lock<std::mutex> lock(lock_);
   for (ScopeImpl* scope : scopes_) {
-    for (auto counter : scope->central_cache_.counters_) {
+    for (auto& counter : scope->central_cache_.counters_) {
       if (names.insert(counter.first).second) {
```

Comment:
would it make any difference in perf, if we use sort and unique on vector rather than using a different set?

Reply:
Hmmm, I'm really not sure.  My intuition says that that solution would be comparable perf at best, but given that there are a lot of variables changing (number of allocations, sorting algorithms, moving elements around in a vector), I think we would need a benchmark to show that this is actually better before moving to a more complicated solution.  Since that's not really the focus of this PR, I'd prefer leaving it as future work.  What do you think?

Reply:
I agree. It is definitely not the focus of this PR. Can we leave a TODO to deal with it later?